src/backends/perf_pt/mod.rs

```diff
@@ -58,6 +57,55 @@ extern "C" {
     fn perf_pt_free_tracer(tr_ctx: *const c_void) -> c_int;
 }
 
+/// A Intel PT trace, obtained via Linux Perf.
+#[allow(dead_code)]
+pub struct PerfPTTrace {
+    buf: *const u8,
+    len: u64,
+}
+
+impl PerfPTTrace {
+    /// Makes a new trace from a raw pointer and a size.
+    ///
+    /// The `buf` argument is assumed to have been allocated on the heap using malloc(3). `len`
```

Comment:
I don't like the fact that the user calls `malloc` but we call `free`. Why don't we just take `len` as an arg and then `malloc` buf ourselves?

Reply:
The user doesn't call malloc, the underlying C code does.

Why are we telling the user about this, is my reflexive question now?! This should probably be mentioned in a non-doc comment somewhere really. Will fix.

Reply:
The doc comment says that the user must have malloc'd the memory that `buf` points to; and we later call `free(buf)` ourselves. Experience suggests that's not a pleasant API. Except in really weird circumstances (which I don't think we have here), it's better to have either the user do both `malloc` & `free` or us do both `malloc` and `free`.

Reply:
The user does neither.

Reply:
OK, let's say "user of this API". I know it's an internal user -- but it's still a user. Is there a reason we can't `malloc` the buffer from inside Rust?

Reply:
That API would be possible, yes, it would have to look like this:

```
let tracer = ...;
let mut trace = tracer::empty_trace();  // does malloc
tracer.start_tracing(&mut trace);
...
tracer.stop_tracing();
... do stuff with trace, buf freed when falls out of scope.
```

We could try this. It's not what we agreed on earlier, but then I guess we didn't think too hard about the consequences of C-level allocations.

One thing I do insist upon though, is that C code is able to `realloc(3)` the buffer. This means rust must be willing for the pointer to move.

Thoughts?

Reply:
To be clear, the user of the API doesn't care currently. The buffer is allocated by `start_tracing` and freed by `drop()`. They never see the memory ops.

Reply:
Couldn't it be something like:

```rust
let tracer = ...;
tracer.start_tracing();
let trace = tracer.stop_tracing();
```

I guess `start_tracing` would say something like "if `buf` is null, `malloc` something" and `stop_tracing` would move the buffer to the trace. `Drop` would make sense on both the trace and tracer in this case, I think.

I admit that I'm not sure where to handle the "C code can realloc" case. Maybe a special method in the tracer `update_buf_ptr` or similar?

Reply:
> I guess start_tracing would say something like "if buf is null, malloc something"

That doesn't quite make sense, but I know roughly what you are getting at [I think]. We could try this but I suggest *not* in this PR.

> I admit that I'm not sure where to handle the "C code can realloc" case. Maybe a special method in the tracer update_buf_ptr or similar?

This can be solved with pointers to pointers.

Reply:
To expand upon that, these lines:

https://github.com/softdevteam/hwtracer/blob/9902d858002e5bd3a4e599f97ea8c7f3a7285122/src/backends/perf_pt/perf_pt.c#L154-L155

Would be replaced with a call to `realloc` which writes to a pointer to a pointer to the allocation. Do you get it?

Reply:
I've raised #33 for this.

Reply:
Oh and #34.

Are you OK for these to be separate PRs? (the realloc one can wait for a while actually).

Reply:
Fine with me.